src/model/helpers.rs

```diff
@@ -694,33 +706,52 @@ pub async fn upsert_subscription_watcher(
     expiry: DateTime<Utc>,
     postgres: &PgPool,
     metrics: Option<&Metrics>,
-) -> Result<(), sqlx::error::Error> {
+) -> Result<(), UpsertSubscriptionWatcherError> {
+    #[derive(Debug, FromRow)]
+    struct Result {
+        // We don't need any fields
+    }
+    let query = "
+        INSERT INTO subscription_watcher (
+            account,
+            project,
+            did_key,
+            sym_key,
+            expiry
+        )
+        SELECT $1, $2, $3, $4, $5 WHERE (
+            SELECT COUNT(*)
+            FROM subscription_watcher
+            WHERE get_address_lower(account)=get_address_lower($1)
+                  AND project=$2
+        ) < $6
+        ON CONFLICT (did_key) DO UPDATE SET
+            updated_at=now(),
+            account=$1,
+            project=$2,
+            sym_key=$4,
+            expiry=$5
+        RETURNING *
+    ";
     let start = Instant::now();
-    let _ = sqlx::query::<Postgres>(
-        "
-            INSERT INTO subscription_watcher (
-                account,
-                project,
-                did_key,
-                sym_key,
-                expiry
-            )
-            VALUES ($1, $2, $3, $4, $5)
-            ON CONFLICT (did_key) DO UPDATE SET
-                updated_at=now(),
-                account=$1,
-                project=$2,
-                sym_key=$4,
-                expiry=$5
-        ",
-    )
-    .bind(account.as_ref())
-    .bind(project)
-    .bind(did_key)
-    .bind(sym_key)
-    .bind(expiry)
-    .execute(postgres)
-    .await?;
+    let mut txn = postgres.begin().await?;
+    // https://stackoverflow.com/a/48730873
+    sqlx::query::<Postgres>("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE") // TODO serialization errors not handled
+        .execute(&mut *txn)
+        .await?;
+    let result = sqlx::query_as::<Postgres, Result>(query)
```

Comment:
why not just an empty tuple if Result is just a bogus placeholder?

Reply:
Good idea, removed in [5b7b480](https://github.com/WalletConnect/notify-server/pull/392/commits/5b7b48065a98deda47f1ad45d67a4ff802aef501)